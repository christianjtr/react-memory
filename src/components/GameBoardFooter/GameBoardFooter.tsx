import React from 'react';
import { useGameContext } from '../../hooks';

const GameBoardFooter = (): React.ReactElement => {
  const {
    state: { score, timer },
  } = useGameContext();

  return (
    <div className="flex my-5 text-xl">
      <div className="w-1/2 font-medium">{`Time: ${timer}`}</div>
      <div className="w-1/2 text-right font-medium">{`Score: ${score}`}</div>
    </div>
  );
};

export default GameBoardFooter;
