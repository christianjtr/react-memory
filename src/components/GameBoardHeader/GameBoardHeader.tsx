import React from 'react';

export interface GameBoardHeaderProps {
  title?: string;
}

const GameBoardHeader: React.FC<GameBoardHeaderProps> = (
  props: GameBoardHeaderProps,
): React.ReactElement => {
  const { title = 'Github Memory' } = props;

  return <h1 className="my-5 text-3xl text-center">{title}</h1>;
};

export default GameBoardHeader;
