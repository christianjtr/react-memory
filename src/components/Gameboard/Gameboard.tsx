import React, { useEffect } from 'react';
import { useGameContext, useGithubContributors } from '../../hooks';
import { GAME_ACTION_TYPES } from '../../contexts/action-types';
import Card from '../Card/Card';

const Gameboard = (): React.ReactElement => {
  const {
    dispatch,
    state: {
      cards,
      score,
      config: { durationInSeconds },
    },
  } = useGameContext();

  const { contributors, fetchData } = useGithubContributors();

  useEffect(() => {
    if (contributors.length > 0) {
      dispatch({ type: GAME_ACTION_TYPES.INIT_GAME, payload: { cards: contributors } });
    }
  }, [contributors]);

  useEffect(() => {
    fetchData();
  }, []);

  return (
    <div className="container mx-auto flex h-screen">
      <div className="w-3/5 m-auto">
        <h1 className="my-5 text-3xl text-center">Github Memory</h1>
        <div className="grid grid-cols-4 gap-4 ">
          {cards.map((card, index) => (
            <Card key={`${card.id}_${index}`} data={card} className="drop-shadow-md" />
          ))}
        </div>
        <div className="flex my-5">
          <div className="w-1/2 font-medium">{`Time: ${durationInSeconds}`}</div>
          <div className="w-1/2 text-right font-medium">{`Score: ${score}`}</div>
        </div>
      </div>
    </div>
  );
};

export default Gameboard;
