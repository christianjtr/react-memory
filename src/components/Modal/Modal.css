.modal {
  @apply fixed z-10 overflow-y-auto top-0 w-full left-0;
}

.modal > .modal__contentainer-wrapper {
  @apply flex items-center justify-center pt-4 px-4 pb-20 text-center sm:block sm:p-0;
}

.modal .modal__center-anchor {
  @apply sm:inline-block align-middle h-screen;
}

.modal .modal__overlay {
  @apply fixed inset-0 transition-opacity;
}

.modal .modal__overlay > .modal__overlay-mask {
  @apply absolute inset-0 bg-gray-900 opacity-75;
}

.modal .modal__container {
  @apply inline-block origin-center bg-white rounded-lg text-left overflow-hidden shadow-xl transform transition-all sm:align-middle sm:max-w-lg w-1/2;
}

.modal .modal__body,
.modal__footer {
  @apply px-4;
}

.modal .modal__body {
  @apply bg-white pt-5 pb-4 sm:p-6 sm:pb-4;
}

.modal .button {
  @apply bg-blue-500 hover:bg-blue-700 text-white py-2 px-4 rounded;
}
