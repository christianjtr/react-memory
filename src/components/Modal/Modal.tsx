import React from 'react';
import './Modal.css';

import { useGameContext } from '../../hooks';

export interface ModalProps {
  onOk: () => void;
}

const Modal: React.FC<ModalProps> = (props: ModalProps): React.ReactElement | null => {
  const { onOk: handleOnOkCallBack } = props;

  const {
    state: { isGameOver, score },
  } = useGameContext();

  if (!isGameOver) return null;

  return (
    <div className="modal" id="modal">
      <div className="modal__contentainer-wrapper">
        <div className="modal__overlay">
          <div className="modal__overlay-mask" />
        </div>
        <span className="modal__center-anchor">&#8203;</span>
        <div
          className="modal__container"
          role="dialog"
          aria-modal="true"
          aria-labelledby="modal-headline">
          <div className="modal__body text-center text-2xl">
            <p>
              <span className="text-center text-5xl">
                GAME
                <br />
                OVER
              </span>
            </p>
            <p className="my-5">{`Score: ${score}`}</p>
            <button
              id="modal-button"
              className="button"
              title="New game!"
              aria-label="New game!"
              onClick={handleOnOkCallBack}>
              New game!
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Modal;
