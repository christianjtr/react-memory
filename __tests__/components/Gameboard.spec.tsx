import React from 'react';
import '@testing-library/jest-dom';
import { render, screen, cleanup } from '@testing-library/react';
import { createRenderer } from 'react-test-renderer/shallow';

import Modal from '../../src/components/Modal/Modal';
import GameBoardHeader from '../../src/components/GameBoardHeader/GameBoardHeader';
import GameBoardFooter from '../../src/components/GameBoardFooter/GameBoardFooter';

import Gameboard from '../../src/components/Gameboard/Gameboard';

expect;

jest.mock('../../src/hooks', () => ({
  useGameContext: jest.fn().mockReturnValue({
    state: {
      cards: [],
    },
  }),
  useGithubContributors: jest.fn().mockResolvedValue({
    contributors: [],
    isLoading: false,
    fetchData: jest.fn().mockResolvedValue([]),
  }),
  useGame: jest.fn().mockReturnValue({
    isValidGameMove: undefined,
    gameCardIds: [],
    GameFunctionalities: {
      addGameCardId: jest.fn(),
      faceUpCard: jest.fn(),
      addToFoundPairs: jest.fn(),
      faceDownCards: jest.fn(),
      initGameMove: jest.fn(),
      startGame: jest.fn(),
    },
  }),
}));

describe('Gameboard component', () => {
  afterEach(() => {
    cleanup();
  });

  it('should render the component correctly', () => {
    const renderer = createRenderer();
    const component = renderer.render(<Gameboard />);

    expect(component).toBeDefined();
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it('should check the children components in shallow mode', () => {
    const renderer = createRenderer();
    renderer.render(<Gameboard />);

    const [firstChildren, modalComponent] = renderer.getRenderOutput().props.children;

    const childrenComponents = [GameBoardHeader, GameBoardFooter];

    childrenComponents.forEach((childComponent) => {
      const child = firstChildren.props.children.find(
        (item: React.ReactElement) => item.type === childComponent,
      );
      expect(child.type).toEqual(childComponent);
    });

    expect(modalComponent.type).toEqual(Modal);
  });
});
