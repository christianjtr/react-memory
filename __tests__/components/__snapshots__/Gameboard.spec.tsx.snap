// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gameboard component should render the component correctly 1`] = `
<div
  className="container mx-auto flex h-screen"
>
  <div
    className="w-4/5 m-auto"
  >
    <GameBoardHeader />
    <div
      className="grid grid-cols-3 md:grid-cols-4 gap-4"
    />
    <GameBoardFooter />
  </div>
  <Modal
    onOk={[Function]}
  />
</div>
`;
